a3/basic-scripting-guide.md

```diff
@@ -0,0 +1,406 @@
+difficult to remember or long-running.
+
+In cases where shell syntax is inappropriate for the task at hand, one can instead call into programs written in other languages, such as Python, which can read from stdin, process data, and write to stdout. 
+
+---
+
+## What's Covered
+* `bash` as a scripting language
+* `python` for system administration
+
+## Scripting with the Bourne-Again Shell (Bash)
+While most programmers are likely familiar with `bash` in its popular capacity as a command line interpreter, it is in fact a powerful and full-featured programming language. Moreover, `bash` has a uniquely qualified claim to the title of *scripting* language in that programs written in `bash` are simply series of shell commands which `bash` reads off and executes line-by-line. Or, conversely, one might say that `bash` command line entries are simply short one-line scripts. So really, you’ve been `bash` scripting all along!
+
+### Shebang!
+Shell scripts typically begin with the **shebang** line: `#!pathtointerpreter`.
+
+[1]: https://en.wikipedia.org/wiki/Shebang_(Unix)#Magic_number
+`#!` is a human-readable representation of a [magic number][1] `0x23 0x21` used by the operating system to determine whether a file is a script or
+an executable binary. If your script is run as an executable e.g. `./awesome_shell_script`, then the operating system will invoke the executable
+(usually an interpreter) at `pathtointerpreter` to run your script. If your script is passed as an argument to an interpreter e.g. 
+`bash awesome_shell_script`, then the shebang has no effect and `bash` will handle the script's execution.
+
+**Why is this important?** The shebang line can be considered a useful piece of metadata which passes the concern of *how* a script is executed from the user to the program's author. `awesome_shell_script` could be a `bash` script, a `python` script, a `ruby` script, etc. The idea is that only the script's behavior, not
+its implementation details, should matter to the user who calls the script.
+
+[2]: http://pubs.opengroup.org/onlinepubs/009695399/utilities/xcu_chap02.html
+You may have seen some variant of `#!/bin/sh`. Although initially referencing the Bourne shell, on modern systems `sh` has come to reference to the
+[Shell Command Language][2], which is a POSIX specification with many implementations including `ash`, `dash`, `ksh88`, etc. `sh` is usually
+symlinked to one of these POSIX-compliant shells. On Debian, for instance, `sh` is symlinked to `dash`. It is important to note that `bash` does **not**
```

Comment:
might want to briefly explain more the difference between bash, ash, dash, etc. There's the discussion of bashisms below, but I think a few more details on the difference (or not mentioning many alts at all) would reduce the potential for confusion.